tests/utests/types/string.c

```diff
@@ -0,0 +1,1327 @@
+    range = ((struct lysc_type_str *)lysc_leaf->type)->length;
+    CHECK_LYSC_RANGE(range, NULL, NULL, NULL, 0, 2, NULL);
+    assert_int_equal(range->parts[0].min_u64, 10);
+    assert_int_equal(range->parts[0].max_u64, 15);
+    assert_int_equal(range->parts[1].min_u64, 255);
+    assert_int_equal(range->parts[1].max_u64, 255);
+    lysp_leaf = (void *) mod->parsed->data;
+    CHECK_LYSP_NODE_LEAF(lysp_leaf, NULL, 0, 0x0, 0, "port", 0, 0, NULL, 0, 0, NULL, NULL);
+    CHECK_LYSP_TYPE(&(lysp_leaf->type), 0, 0, 0, 0, 0, 0x10, 0, 1, "my_type", 0, 0, 1, 0, 0, 0);
+    CHECK_LYSP_RESTR(lysp_leaf->type.length, "min .. 15 | max", NULL, NULL, NULL, 0, NULL);
+
+    // ERROR TESTS NEGATIVE VALUE
+    schema = MODULE_CREATE_YANG("ERR0", "leaf port {type string {length \"-1 .. 20\";}}");
+    UTEST_INVALID_MODULE(schema, LYS_IN_YANG, NULL, LY_EVALID);
+    CHECK_LOG_CTX("Invalid length restriction - value \"-1\" does not fit the type limitations.",
+            "/ERR0:port");
+
+    schema = MODULE_CREATE_YANG("ERR1", "leaf port {type string {length \"100 .. 18446744073709551616\";}}");
+    UTEST_INVALID_MODULE(schema, LYS_IN_YANG, NULL, LY_EVALID);
+    CHECK_LOG_CTX("Invalid length restriction - invalid value \"18446744073709551616\".",
+            "/ERR1:port");
+
+    schema = MODULE_CREATE_YANG("ERR2", "leaf port {type string {length \"10 .. 20 | 20 .. 30\";}}");
+    UTEST_INVALID_MODULE(schema, LYS_IN_YANG, NULL, LY_EVALID);
+    CHECK_LOG_CTX("Invalid length restriction - values are not in ascending order (20).",
+            "/ERR2:port");
+
+    schema = MODULE_CREATE_YANG("ERR3",
+            "typedef my_type {"
+            "    type string;"
+            "}"
+            "leaf port {type my_type {length \"-1 .. 15\";}}");
+    UTEST_INVALID_MODULE(schema, LYS_IN_YANG, NULL, LY_EVALID);
+    CHECK_LOG_CTX("Invalid length restriction - value \"-1\" does not fit the type limitations.",
+            "/ERR3:port");
+
+    /////////////////////////////
+    // PATTERN
+    schema = MODULE_CREATE_YANG("TPATTERN_0", "leaf port {type string"
+            "{pattern '[a-zA-Z_][a-zA-Z0-9\\-_.]*';}"
+            "}");
+    UTEST_ADD_MODULE(schema, LYS_IN_YANG, NULL, &mod);
+    assert_non_null(mod);
+    lysc_leaf = (void *) mod->compiled->data;
+    CHECK_LYSC_NODE_LEAF(lysc_leaf, NULL, 0, 0x5, 1, "port", 0, 0, 0, NULL, 0, 0, NULL, NULL);
+    CHECK_LYSC_TYPE_STR((struct lysc_type_str *)lysc_leaf->type, 0, 0, 1);
+    pattern = ((struct lysc_type_str *)lysc_leaf->type)->patterns[0];
+    CHECK_LYSC_PATTERN(pattern, NULL, NULL, NULL, "[a-zA-Z_][a-zA-Z0-9\\-_.]*", 0, 0, NULL);
+    lysp_leaf = (void *) mod->parsed->data;
+    CHECK_LYSP_NODE_LEAF(lysp_leaf, NULL, 0, 0x0, 0, "port", 0, 0, NULL, 0, 0, NULL, NULL);
+    CHECK_LYSP_TYPE(&(lysp_leaf->type), 0, 0, 0, 0, 0, 0x40, 0, 0, "string", 0, 1, 1, 0, 0, 0);
+    // CHECK_LYSP_RESTR(&(lysp_leaf->type.patterns[0]), "[a-zA-Z_][a-zA-Z0-9\\-_.]*", NULL, NULL, NULL, 0, NULL);
```

Comment:
seems that pattern checking is not covered, is it correct or just forgotten comments?